File: packages/plugin-sdk/scripts/test-entrypoints-codec.mjs
// Round-trip Cap'n codecs for nullable Entrypoints interface fields.
// Generated write must skip exportCap(null); read must omit null caps.
import assert from "node:assert/strict";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const root = dirname(fileURLToPath(import.meta.url));
const dist = join(root, "../dist");

const { EntrypointsCodec, encodeMessage, decodeMessage } = await import(
  join(dist, "generated-wire.js")
);

function capTable() {
  /** @type {unknown[]} */
  const slots = [];
  return {
    exportCap(value) {
      if (typeof value !== "string" || !value) {
        throw new Error("exportCap requires a non-empty object id");
      }
      const i = slots.length;
      slots.push(value);
      return i;
    },
    importCap(index) {
      if (index === null || index === undefined) return null;
      return slots[index] ?? null;
    },
  };
}

function roundTrip(value) {
  const caps = capTable();
  const bytes = encodeMessage(EntrypointsCodec, value, caps);
  return decodeMessage(EntrypointsCodec, bytes, caps);
}

function assertAbsent(decoded, names) {
  for (const name of names) {
    assert.equal(decoded[name], undefined, `${name} must stay absent`);
  }
}

const ALL = [
  "eventConsumer",
  "jobRunner",
  "storefront",
  "storage",
  "databaseAdapter",
  "remoteLibrary",
  "cli",
  "oidc",
];

{
  const decoded = roundTrip({});
  assertAbsent(decoded, ALL);
}

{
  const decoded = roundTrip({ storefront: "sf" });
  assert.equal(decoded.storefront, "sf");
  assertAbsent(
    decoded,
    ALL.filter((n) => n !== "storefront"),
  );
}

{
  const decoded = roundTrip({
    storefront: "sf",
    cli: "cli",
    eventConsumer: "ev",
  });
  assert.equal(decoded.storefront, "sf");
  assert.equal(decoded.cli, "cli");
  assert.equal(decoded.eventConsumer, "ev");
  assertAbsent(decoded, ["jobRunner", "storage", "databaseAdapter", "remoteLibrary", "oidc"]);
}

{
  const all = Object.fromEntries(ALL.map((n) => [n, n]));
  const decoded = roundTrip(all);
  for (const name of ALL) {
    assert.equal(decoded[name], name);
  }
}

{
  const decoded = roundTrip({ eventConsumer: "ev", jobRunner: "job" });
  assert.equal(decoded.eventConsumer, "ev");
  assert.equal(decoded.jobRunner, "job");
  assertAbsent(
    decoded,
    ALL.filter((n) => n !== "eventConsumer" && n !== "jobRunner"),
  );
}

{
  const decoded = roundTrip({ cli: "cli-only" });
  assert.equal(decoded.cli, "cli-only");
  assertAbsent(
    decoded,
    ALL.filter((n) => n !== "cli"),
  );
}

{
  const caps = capTable();
  assert.doesNotThrow(() => encodeMessage(EntrypointsCodec, { storefront: null }, caps));
  const decoded = decodeMessage(
    EntrypointsCodec,
    encodeMessage(EntrypointsCodec, { storefront: null }, capTable()),
    capTable(),
  );
  assert.equal(decoded.storefront, undefined);
}

console.log("ok: Entrypoints nullable interface codec round-trips");

File: packages/plugin-sdk-python/tests/test_entrypoints_codec.py
"""Nullable Entrypoints interface fields must not call export_cap(None)."""

from __future__ import annotations

import unittest

from bookclerk_plugin_sdk import _wire as wire


class _Caps:
    def __init__(self) -> None:
        self.slots: list[object] = []

    def export_cap(self, value: object) -> int:
        if value is None:
            raise AssertionError("export_cap must not be called with None")
        idx = len(self.slots)
        self.slots.append(value)
        return idx

    def import_cap(self, index: int | None) -> object | None:
        if index is None:
            return None
        return self.slots[index]


ALL = (
    "eventConsumer",
    "jobRunner",
    "storefront",
    "storage",
    "databaseAdapter",
    "remoteLibrary",
    "cli",
    "oidc",
)


def _round_trip(value: dict) -> dict:
    caps = _Caps()
    data = wire._encode_message(wire._entrypoints_codec, value, caps)
    return wire._decode_message(wire._entrypoints_codec, data, caps)


class EntrypointsNullCapTests(unittest.TestCase):
    def test_no_optional_entrypoints(self) -> None:
        decoded = _round_trip({})
        for name in ALL:
            self.assertNotIn(name, decoded)

    def test_one_entrypoint(self) -> None:
        decoded = _round_trip({"storefront": "sf"})
        self.assertEqual(decoded["storefront"], "sf")
        for name in ALL:
            if name != "storefront":
                self.assertNotIn(name, decoded)

    def test_several_entrypoints(self) -> None:
        decoded = _round_trip(
            {"storefront": "sf", "cli": "cli", "eventConsumer": "ev"}
        )
        self.assertEqual(decoded["storefront"], "sf")
        self.assertEqual(decoded["cli"], "cli")
        self.assertEqual(decoded["eventConsumer"], "ev")
        self.assertNotIn("jobRunner", decoded)

    def test_all_entrypoints(self) -> None:
        payload = {name: name for name in ALL}
        decoded = _round_trip(payload)
        self.assertEqual(decoded, payload)

    def test_event_job_only(self) -> None:
        decoded = _round_trip({"eventConsumer": "ev", "jobRunner": "job"})
        self.assertEqual(decoded["eventConsumer"], "ev")
        self.assertEqual(decoded["jobRunner"], "job")
        self.assertNotIn("storefront", decoded)

    def test_named_entrypoint_only(self) -> None:
        decoded = _round_trip({"cli": "cli-only"})
        self.assertEqual(decoded["cli"], "cli-only")
        self.assertNotIn("storefront", decoded)

    def test_explicit_none_does_not_export(self) -> None:
        decoded = _round_trip({"storefront": None})
        self.assertNotIn("storefront", decoded)


if __name__ == "__main__":
    unittest.main()
